fdbserver/GrvProxyServer.actor.cpp

```diff
@@ -302,6 +304,125 @@ ACTOR Future<Void> healthMetricsRequestServer(GrvProxyInterface grvProxy,
+					tr->set(GlobalConfig::prefixedKey(fdbClientInfoTxnSampleRate), rate.pack());
+				}
+				if (sizeLimit.present()) {
+					const int64_t sizeLimitInt =
+					    BinaryReader::fromStringRef<int64_t>(sizeLimit.get().contents(), Unversioned());
+					Tuple size = Tuple().append(sizeLimitInt);
+					tr->set(GlobalConfig::prefixedKey(fdbClientInfoTxnSizeLimit), size.pack());
+				}
+
+				wait(tr->commit());
+				break;
+			} catch (Error& e) {
+				// Multilpe GRV proxies may attempt this migration at the same
+				// time, sometimes resulting in aborts due to conflicts.
+				// Purposefully avoid retrying, making this migration
+				// best-effort.
+				TraceEvent(SevInfo, "GlobalConfigRetryableMigrationError").errorUnsuppressed(e).suppressFor(1.0);
+				wait(tr->onError(e));
+			}
+		}
+	} catch (Error& e) {
+		// Catch non-retryable errors (and do nothing).
+		TraceEvent(SevWarnAlways, "GlobalConfigMigrationError").error(e);
+	}
+	return Void();
+}
+
+ACTOR Future<Void> globalConfigHandleRefreshBatch(std::vector<GlobalConfigRefreshRequest> requests,
+                                                  GrvProxyData* grvProxyData) {
+	if (requests.size() == 0) {
+		return Void();
+	}
+
+	state Reference<ReadYourWritesTransaction> tr = makeReference<ReadYourWritesTransaction>(grvProxyData->cx);
+	loop {
+		try {
+			tr->setOption(FDBTransactionOptions::READ_SYSTEM_KEYS);
+			RangeResult result = wait(tr->getRange(globalConfigDataKeys, CLIENT_KNOBS->TOO_MANY));
+			for (const auto& req : requests) {
+				req.reply.send(GlobalConfigRefreshReply{ result.arena(), result });
+			}
+			break;
+		} catch (Error& e) {
+			wait(tr->onError(e));
+		}
+	}
+	return Void();
+}
+
+ACTOR Future<Void> globalConfigRefreshBatcher(GrvProxyData* grvProxyData,
+                                              GrvProxyInterface grvProxy,
+                                              PromiseStream<std::vector<GlobalConfigRefreshRequest>> out) {
+	loop {
+		state Future<Void> timeout = delay(SERVER_KNOBS->GLOBAL_CONFIG_MAX_REFRESH_BATCH_INTERVAL);
+		state std::vector<GlobalConfigRefreshRequest> requests;
+		loop {
+			choose {
+				when(GlobalConfigRefreshRequest refresh = waitNext(grvProxy.refreshGlobalConfig.getFuture())) {
+					requests.push_back(refresh);
+				}
+				when(wait(timeout)) { break; }
+			}
+		}
+		if (requests.size() > 0) {
+			out.send(requests);
+		}
+	}
+}
+
+// Handle common GlobalConfig transactions on the server side, because not all
+// clients are allowed to read system keys. Eventually, this could become its
+// own role.
+ACTOR Future<Void> globalConfigRequestServer(GrvProxyData* grvProxyData, GrvProxyInterface grvProxy) {
+	state ActorCollection actors(false);
+	state PromiseStream<std::vector<GlobalConfigRefreshRequest>> batchedRequests;
+	actors.add(globalConfigRefreshBatcher(grvProxyData, grvProxy, batchedRequests));
+
+	// Run one-time migration supporting upgrades.
+	wait(success(globalConfigMigrate(grvProxyData)));
```

Comment:
@sfc-gh-ajbeamon The other alternative here is to run the migration without blocking. But since the migration should complete before refresh requests are handled, I think this solution of accepting requests and  batching them, but not handling the batched requests until the migration is complete makes sense.

Reply:
I think this is similar to the old behavior where clients would first migrate and then refresh, so that sounds fine.

Reply:
On the topic of accumulating requests, I guess we could also have that problem here if the migration never completes. In that case, we might either need some way to guard our memory and/or we could timeout the migration attempt and then subsequently flush the requests.

Reply:
Adding a timeout to this call will be ok I think, this migration should work but it shouldn't block other things if it gets stuck.